### PR TITLE
ovirt: Add survey

### PR DESCRIPTION
**Pull request by rgolangh**
- move ovirt from `HiddenPlatformNames` to `PlatformsNames`    
- ask for credentials      
- ask for a cluster   
- ask for a storage domain
- ask for a network     
- recommend not to use default network

### CODE REVIEW

**Comment by rgolangh** [comment]
@patrickdillon  PTAL

File: pkg/asset/installconfig/ovirt/ovirt.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func askCredentials() (ovirt.OvirtConfig, error) {
+	c := ovirt.OvirtConfig{}
+	err := survey.Ask([]*survey.Question{
+		{
+			Prompt: &survey.Input{
+				Message: "Enter oVirt's api endpoint URL",
+				Help:    "oVirt engine api url",
+				Default: "https://ovirt-engine-fqdn/ovirt-engine/api",
```
Comment (patrickdillon) [inline pkg/asset/installconfig/ovirt/ovirt.go]: is this a valid default value? it looks like an example, in which case I don't think it should be a default and should perhaps be moved to `Message`

File: pkg/asset/installconfig/ovirt/ovirt.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+			Validate: survey.ComposeValidators(survey.Required),
+		},
+	}, &c.URL)
+
+	err = survey.Ask([]*survey.Question{
+		{
+			Prompt: &survey.Input{
+				Message: "Enter ovirt-engine username",
+				Help:    "The user must have permissions to create VMs, Disks, on the Storage Domain with the name matching the openshift cluster",
+				Default: "admin@internal",
```
Comment (patrickdillon) [inline pkg/asset/installconfig/ovirt/ovirt.go]: is this a valid default value? it looks like an example, in which case I don't think it should be a default and should perhaps be moved to Message

File: pkg/types/ovirt/config.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	"path/filepath"
+
+	"gopkg.in/yaml.v2"
+)
+
+var defaultOvirtConfigEnvVar = "OVIRT_CONFIG"
+var defaultOvirtConfigPath = filepath.Join(os.Getenv("HOME"), ".ovirt", "ovirt-config.yaml")
+
+// OvirtConfig holds oVirt api access details
+type OvirtConfig struct {
```
Comment (patrickdillon) [inline pkg/types/ovirt/config.go]: This type is duplicated elsewhere: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/ovirt/config.go#L19

This (pkg types) is a better place for it. Would be nice to centralize this as part of this pr.

Comment (rgolangh) [reply pkg/types/ovirt/config.go]: good catch, done.

File: pkg/asset/installconfig/ovirt/ovirt.go (hunk 2 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/ovirt.go]: Its a proper default value for the platform. It is recommended to create a dedicated user
for this setup, and I will add that to the help. In reality ovirt comes with a single, predefined super user.

File: pkg/asset/installconfig/ovirt/ovirt.go (hunk 1 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/ovirt.go]: good call, changing.

File: pkg/asset/installconfig/ovirt/cluster.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	if !ok {
+		return "", fmt.Errorf("there are no available clusters")
+	}
+
+	for _, cluster := range clusters.Slice() {
+		clusterByNames[cluster.MustName()] = cluster
+		clusterNames = append(clusterNames, cluster.MustName())
+	}
+	err = survey.AskOne(&survey.Select{
+		Message: "Pick the oVirt cluster",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/cluster.go]: Select the oVirt cluster

File: pkg/asset/installconfig/ovirt/cluster.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		Help:    "The oVirt cluster under which the VMs will be created.",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/cluster.go]: The oVirt cluster where the VMs will be created.

File: pkg/asset/installconfig/ovirt/credentials.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return c, err
+	}
+	c.Insecure = !ovirtCertTrusted
+
+	err = survey.Ask([]*survey.Question{
+		{
+			Prompt: &survey.Input{
+				Message: "Enter ovirt-engine username",
+				Help:    "The user must have permissions to create VMs, Disks, on the Storage Domain with the name matching the Openshift cluster",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/credentials.go]: The user must have permissions to create VMs and disks on the Storage Domain with the same name as the Openshift cluster",

File: pkg/asset/installconfig/ovirt/ovirt.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	}, &p.DNSVIP)
+	if err != nil {
+		return nil, err
+	}
+
+	err = survey.Ask([]*survey.Question{
+		{
+			Prompt: &survey.Input{
+				Message: "Enter the ingress IP ",
+				Help:    "Make sure the ip is not used by any party",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/ovirt.go]: Make sure the IP address is not in use

File: pkg/asset/installconfig/ovirt/validators.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+		if err != nil {
+			return errors.Errorf("failed to construct connection to oVirt platform %s", err)
+		}
+
+		defer connection.Close()
+
+		err = connection.Test()
+		if err != nil {
+			return errors.Errorf("connection to oVirt platform test failed %s", err)
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/validators.go]: Failed to connect to oVirt platform

File: pkg/asset/installconfig/ovirt/ovirt.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	}, &p.APIVIP)
+	if err != nil {
+		return nil, err
+	}
+
+	err = survey.Ask([]*survey.Question{
+		{
+			Prompt: &survey.Input{
+				Message: "Enter the internal DNS Virtual IP",
+				Help:    "Make sure the ip is not used by any party",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/ovirt.go]: Make sure the IP address is not in use

File: pkg/asset/installconfig/ovirt/ovirt.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	err = askNetwork(c, &p)
+	if err != nil {
+		return &p, err
+	}
+
+	err = survey.Ask([]*survey.Question{
+		{
+			Prompt: &survey.Input{
+				Message: "Enter the internal API Virtual IP",
+				Help:    "Make sure the ip is not used by any party",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/ovirt.go]: Make sure the IP address is not in use

File: pkg/asset/installconfig/ovirt/storage.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+	if !ok {
+		return fmt.Errorf("there are no available storage domains for cluster %s", clusterName)
+	}
+
+	for _, domain := range domains.Slice() {
+		domainsForCluster[domain.MustName()] = domain
+		domainNames = append(domainNames, domain.MustName())
+	}
+	err = survey.AskOne(&survey.Select{
+		Message: "Pick the oVirt storage domain",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/storage.go]: Select the oVirt storage domain

File: pkg/asset/installconfig/ovirt/network.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+	if !ok {
+		return fmt.Errorf("there are no available networks for cluster %s", p.ClusterID)
+	}
+
+	for _, network := range networks.Slice() {
+		networkByNames[network.MustName()] = network
+		networkNames = append(networkNames, network.MustName())
+	}
+	err = survey.AskOne(&survey.Select{
+		Message: "Pick the oVirt network",
```
Comment (peterclauterbach) [inline pkg/asset/installconfig/ovirt/network.go]: Select the oVirt network

File: pkg/asset/installconfig/ovirt/cluster.go (hunk 4 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/cluster.go]: done

File: pkg/asset/installconfig/ovirt/cluster.go (hunk 5 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/cluster.go]: done

File: pkg/asset/installconfig/ovirt/credentials.go (hunk 6 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/credentials.go]: done

File: pkg/asset/installconfig/ovirt/ovirt.go (hunk 7 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/ovirt.go]: done

File: pkg/asset/installconfig/ovirt/validators.go (hunk 8 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/validators.go]: done

File: pkg/asset/installconfig/ovirt/ovirt.go (hunk 9 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/ovirt.go]: done

File: pkg/asset/installconfig/ovirt/ovirt.go (hunk 10 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/ovirt.go]: done

File: pkg/asset/installconfig/ovirt/storage.go (hunk 11 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/storage.go]: done

File: pkg/asset/installconfig/ovirt/network.go (hunk 12 above)
Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/network.go]: done

File: pkg/types/ovirt/client.go [hunk 13]
```diff
@@ -6,13 +6,13 @@ import (
 
 // GetConnection is a convenience method to get a connection to ovirt api
 // form a Config Object.
-func GetConnection(config Config) (*ovirtsdk.Connection, error) {
+func GetConnection(ovirtConfig Config) (*ovirtsdk.Connection, error) {
```
Comment (abhinavdahiya) [inline pkg/types/ovirt/client.go]: `pkg/types` is API package for installer, so the consumers of it shouldn't force to vendor in SDKs...

the `pkg/types` should probably not have ovirt client dependencies..
is there is reason why these are moved from asset/installconfig to pkg/types

Comment (patrickdillon) [reply pkg/types/ovirt/client.go]: > is there is reason why these are moved from asset/installconfig to pkg/types
It was in response to: https://github.com/openshift/installer/pull/3047#discussion_r374751508

I think a proper solution would be to just move the type and keep the rest of the logic in asset.

File: pkg/asset/installconfig/ovirt/network.go [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	for _, network := range networks.Slice() {
+		networkByNames[network.MustName()] = network
+		networkNames = append(networkNames, network.MustName())
+	}
+	err = survey.AskOne(&survey.Select{
+		Message: "Select the oVirt network",
+		Help: "The oVirt network of the deployed VMs. 'ovirtmgmt' is the default network - it is recommended " +
+			"to work with a dedicated network per Openshift cluster",
```
Comment (crawford) [inline pkg/asset/installconfig/ovirt/network.go]: OpenShift

File: pkg/types/ovirt/client.go (hunk 13 above)
Comment (rgolangh) [reply pkg/types/ovirt/client.go]: I'm moving it back the original `pkg/asset/installconfig`

File: pkg/asset/installconfig/ovirt/credentials.go [hunk 15]
```diff
@@ -0,0 +1,66 @@
+package ovirt
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/ovirt/credentials.go]: I didn't realize that the install-config yaml for ovirt has the credentials inlined.. 
This is not safe :( 
we should be using the env based sources for fetching the creds :(

Comment (patrickdillon) [reply pkg/asset/installconfig/ovirt/credentials.go]: fwiw we discussed this before to some extent: https://github.com/openshift/installer/pull/1948#discussion_r339841555

Comment (abhinavdahiya) [reply pkg/asset/installconfig/ovirt/credentials.go]: Yeah, we should have implemented your proposal before merging, but i wouldn't block this PR on this change now :(

**Comment by abhinavdahiya** [comment]
/hold

to stop the retest bot, otherwise it will keep retesting. please remove this when there is consensus this can be merged in feature freeze.

**Comment by crawford** [comment]
The oVirt failures are due to https://bugzilla.redhat.com/show_bug.cgi?id=1801898, which is orthogonal to this PR. This is important enough that I’m going to merge it despite the CI issue.

**Comment by crawford** [comment]
Couldn’t connect to the internal registry while building images.

/retest

**Pull request merged**